### ISSUE TITLE
Engine crash when GDScript "is" keyword is tested against a variable containing a string

### ISSUE THREAD

**Issue by nsrosenqvist** [godotengine/godot#55547]
### Godot version

3.4.stable.flathub.206ba70f4

### System information

Ubuntu 21.04 x64, Intel(R) Core(TM) i5-3317U CPU @ 1.70GHz, Mesa 21.0.3, GLES3

### Issue description

The GDScript _is_ keyword is supposed to be tested against either a native class name or GDScript instance. If, however, a variable would hold something other than a reference to either of these then the engine crashes silently. This is the resulting stack trace:

```
Running: /app/bin/godot-bin --path /home/niklas/Projects/TestProject --remote-debug 127.0.0.1:6007 --allow_focus_steal_pid 17 --position 288,150
Godot Engine v3.4.stable.flathub.206ba70f4 - https://godotengine.org
OpenGL ES 3.0 Renderer: Mesa Intel(R) HD Graphics 4000 (IVB GT2)
OpenGL ES Batching: ON
 
Doesn't crash if code below is commented out.

================================================================
handle_crash: Program crashed with signal 11
Engine version: Godot Engine v3.4.stable.flathub (206ba70f4b2a3eac4455bb21bd7c8af76285bd86)
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] /usr/lib/x86_64-linux-gnu/libc.so.6(+0x39530) [0x7f37b3a75530] ()
[2] /app/bin/godot-bin() [0xabb0c0] ()
[3] /app/bin/godot-bin() [0xa3beb5] ()
[4] /app/bin/godot-bin() [0x9fe449] ()
[5] /app/bin/godot-bin() [0x1cec9e8] ()
[6] /app/bin/godot-bin() [0x1cec95d] ()
[7] /app/bin/godot-bin() [0x1d1bca7] ()
[8] /app/bin/godot-bin() [0x1d38ebf] ()
[9] /app/bin/godot-bin() [0x94d19a] ()
[10] /usr/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0) [0x7f37b3a5fb90] ()
[11] /app/bin/godot-bin() [0x9612ce] ()
-- END OF BACKTRACE --
================================================================
```

### Steps to reproduce

I've been able to reproduce this consistently using the code included under "minimal reproduction project", simply attach it to a node and it should trigger the crash. The first _print_ call is visible in the terminal but not in the editor output pane.

### Minimal reproduction project

```gdscript
extends Spatial

func _ready():
	var test_class = Spatial

	if self is test_class:
		print("Doesn't crash if code below is commented out.")
	
	# If this below is uncommented then the code above doesn't even run.
	test_class = "Spatial"

	if self is test_class:
		print("Crashes without a message and no warning in editor.")

```

**Comment by nsrosenqvist** [comment]
Sorry about that! Thanks for confirming the issue and providing one!

**Comment by Rubonnek** [comment]
The issue comes from casting a variant pointer to an `Object` pointer.

The faulty line is at:

https://github.com/godotengine/godot/blob/eef08ea3e06a3c4ca6fefb63a7bab113b42a3265/modules/gdscript/gdscript_function.cpp#L485

`obj_B` is a null pointer that later gets derefenced at:

https://github.com/godotengine/godot/blob/eef08ea3e06a3c4ca6fefb63a7bab113b42a3265/modules/gdscript/gdscript_function.cpp#L512

causing the crash if `DEBUG_ENABLED` is true.

Otherwise, `obj_B` will make the `nc` pointer null at line:
https://github.com/godotengine/godot/blob/eef08ea3e06a3c4ca6fefb63a7bab113b42a3265/modules/gdscript/gdscript_function.cpp#L508

which will get dereferenced at line:
https://github.com/godotengine/godot/blob/eef08ea3e06a3c4ca6fefb63a7bab113b42a3265/modules/gdscript/gdscript_function.cpp#L516

causing a crash due to a null pointer dereference.

**Comment by akien-mga** [comment]
Fixed by #56791.
